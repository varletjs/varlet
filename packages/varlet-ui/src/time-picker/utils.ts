import { hours24, hoursAmpm } from './props'
import type { AmPm, AllowedTime } from './props'
import { toNumber } from '@varlet/shared'

type DisableProps = {
  time: number
  format: string
  ampm: AmPm | undefined
  hour: string
  minute: number
  max: string | undefined
  min: string | undefined
  disableHour: Array<string>
  allowedTime: AllowedTime | undefined
}

export const notConvert = (format: string, ampm: AmPm | undefined): boolean => format === '24hr' || ampm === 'am'

export const convertHour = (format: string, ampm: AmPm | undefined, hour: string) => {
  const index = hoursAmpm.findIndex((hourAmpm) => toNumber(hourAmpm) === toNumber(hour))
  const getHour = notConvert(format, ampm) ? hour : hours24[index]

  return {
    hourStr: getHour,
    hourNum: toNumber(getHour),
  }
}

export const getNumberTime = (time: string) => {
  const [hour, minute, second] = time.split(':')

  return {
    hour: toNumber(hour),
    minute: toNumber(minute),
    second: toNumber(second),
  }
}

export const getIsDisableMinute = (values: Omit<DisableProps, 'minute'>): boolean => {
  const { time, format, ampm, hour, max, min, disableHour } = values
  const { hourStr, hourNum } = convertHour(format, ampm, hour)
  let isBetweenMinMax = false
  let isAllow = false

  if (disableHour.includes(hourStr)) return true

  if (max && !min) {
    const { hour: maxHour, minute: maxMinute } = getNumberTime(max)
    isBetweenMinMax = maxHour === hourNum && time > maxMinute
  }

  if (!max && min) {
    const { hour: minHour, minute: minMinute } = getNumberTime(min)

    isBetweenMinMax = minHour === hourNum && time < minMinute
  }

  if (max && min) {
    const { hour: maxHour, minute: maxMinute } = getNumberTime(max)
    const { hour: minHour, minute: minMinute } = getNumberTime(min)

    isBetweenMinMax = (minHour === hourNum && time < minMinute) || (maxHour === hourNum && time > maxMinute)
  }

  if (values.allowedTime?.minutes) isAllow = values.allowedTime?.minutes(time)

  return isBetweenMinMax || isAllow
}

export const getIsDisableSecond = (values: DisableProps): boolean => {
  const { time, format, ampm, hour, minute, max, min, disableHour } = values
  const { hourStr, hourNum } = convertHour(format, ampm, hour)
  let isBetweenMinMax = false
  let isAllow = false

  if (disableHour.includes(hourStr)) return true

  if (max && !min) {
    const { hour: maxHour, minute: maxMinute, second: maxSecond } = getNumberTime(max)

    isBetweenMinMax = (maxHour === hourNum && maxMinute < minute) || (maxMinute === minute && time > maxSecond)
  }

  if (!max && min) {
    const { hour: minHour, minute: minMinute, second: minSecond } = getNumberTime(min)

    isBetweenMinMax = (minHour === hourNum && minMinute > minute) || (minMinute === minute && time > minSecond)
  }

  if (max && min) {
    const { hour: maxHour, minute: maxMinute, second: maxSecond } = getNumberTime(max)
    const { hour: minHour, minute: minMinute, second: minSecond } = getNumberTime(min)

    isBetweenMinMax =
      (maxHour === hourNum && maxMinute < minute) ||
      (minHour === hourNum && minMinute > minute) ||
      (maxHour === hourNum && maxMinute === minute && time > maxSecond) ||
      (minHour === hourNum && minMinute === minute && time < minSecond)
  }

  if (values.allowedTime?.seconds) isAllow = values.allowedTime?.seconds(time)

  return isBetweenMinMax || isAllow
}
