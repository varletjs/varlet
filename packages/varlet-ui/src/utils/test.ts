import { nextTick } from 'vue'
import type { ComponentPublicInstance } from 'vue'
import { config } from '@vue/test-utils'
import type { VueWrapper, DOMWrapper } from '@vue/test-utils'
import { isPlainObject } from '@varlet/shared'

export const delay = (time: number) => new Promise((resolve) => setTimeout(resolve, time))

export function getTouch(el: Element | Window, x: number, y: number) {
  return {
    identifier: Date.now(),
    target: el,
    pageX: x,
    pageY: y,
    clientX: x,
    clientY: y,
    radiusX: 2.5,
    radiusY: 2.5,
    rotationAngle: 10,
    force: 0.5,
  }
}

export function mockDoubleRaf() {
  const originMethod = window.requestAnimationFrame

  Object.assign(window, {
    requestAnimationFrame(fn: any) {
      setTimeout(fn, 16)
    },
  })

  return {
    mockRestore() {
      window.requestAnimationFrame = originMethod
    },
  }
}

export function trigger(
  wrapper: VueWrapper<ComponentPublicInstance<any, any, any>> | DOMWrapper<Element> | Element | Window,
  eventName: string,
  x = 0,
  y = 0,
  offsetX = 0,
  offsetY = 0
) {
  const el = 'element' in wrapper ? wrapper.element : wrapper
  const touchList = [getTouch(el, x, y)]

  const event = new CustomEvent(eventName, { bubbles: true, cancelable: true, detail: {} })

  Object.assign(event, {
    clientX: x,
    clientY: y,
    offsetX,
    offsetY,
    touches: touchList,
    targetTouches: touchList,
    changedTouches: touchList,
  })

  el.dispatchEvent(event)

  return nextTick()
}

export function mockOffset({
  offsetWidth,
  offsetHeight,
  clientWidth,
  clientHeight,
  offsetLeft,
  offsetTop,
}: {
  offsetWidth?: number
  offsetHeight?: number
  clientWidth?: number
  clientHeight?: number
  offsetLeft?: number
  offsetTop?: number
} = {}) {
  Object.defineProperties(HTMLElement.prototype, {
    offsetWidth: {
      get() {
        return offsetWidth ?? (parseFloat(window.getComputedStyle(this).width) || 0)
      },
    },
    offsetHeight: {
      get() {
        return offsetHeight ?? (parseFloat(window.getComputedStyle(this).height) || 0)
      },
    },
    clientWidth: {
      get() {
        return clientWidth ?? (parseFloat(window.getComputedStyle(this).width) || 0)
      },
    },
    clientHeight: {
      get() {
        return clientHeight ?? (parseFloat(window.getComputedStyle(this).height) || 0)
      },
    },
    offsetLeft: {
      get() {
        return offsetLeft ?? (parseFloat(window.getComputedStyle(this).marginLeft) || 0)
      },
    },
    offsetTop: {
      get() {
        return offsetTop ?? (parseFloat(window.getComputedStyle(this).marginTop) || 0)
      },
    },
    offsetParent: {
      get() {
        return this.parentNode ?? {}
      },
    },
  })
}

export function mockImageNaturalSize(naturalWidth: number, naturalHeight: number) {
  Object.defineProperties(HTMLImageElement.prototype, {
    naturalWidth: {
      get() {
        return naturalWidth
      },
    },
    naturalHeight: {
      get() {
        return naturalHeight
      },
    },
  })
}

export async function triggerDrag(
  el: VueWrapper<ComponentPublicInstance<any, any, any>> | DOMWrapper<Element> | HTMLElement,
  x = 0,
  y = 0
) {
  await trigger(el, 'touchstart', 0, 0)
  await trigger(el, 'touchmove', x / 4, y / 4)
  await trigger(el, 'touchmove', x / 3, y / 3)
  await trigger(el, 'touchmove', x / 2, y / 2)
  await trigger(el, 'touchmove', x, y)
  await trigger(el, 'touchend', x, y)
}

export function mockTranslate() {
  const originMethod = window.getComputedStyle

  const XRE = /translateX\((\d+)px\)/
  const YRE = /translateY\((\d+)px\)/

  Object.assign(window, {
    getComputedStyle: (el: HTMLElement) => {
      const styles = originMethod.call(window, el)
      const x = styles.transform.match(XRE)?.[1] ?? 0
      const y = styles.transform.match(YRE)?.[1] ?? 0
      styles.transform = `matrix(1, 0, 0, 1, ${x}, ${y})`
      return styles
    },
  })

  return {
    mockRestore() {
      window.getComputedStyle = originMethod
    },
  }
}

export function mockFileReader(url: string) {
  const originMethod = window.FileReader

  class FileReader {
    result = ''

    onload() {}

    readAsDataURL() {
      this.result = url
      this.onload()
    }
  }

  Object.assign(window, {
    FileReader,
  })

  return {
    mockRestore() {
      window.FileReader = originMethod
    },
  }
}

export function mockStubs() {
  const originStubs = config.global.stubs

  config.global.stubs = {}

  return {
    mockRestore() {
      config.global.stubs = originStubs
    },
  }
}

export function mockConsole(method: keyof Console, fn: any = () => {}) {
  const originMethod = console[method]

  console[method] = fn

  return {
    mockRestore() {
      console[method] = originMethod as any
    },
  }
}

export function mockScrollTo(Element: any) {
  Element.prototype.scrollTo = function (x: ScrollToOptions | number, y?: number) {
    if (isPlainObject(x)) {
      this.scrollLeft = x.left
      this.scrollTop = x.top
    } else {
      this.scrollLeft = x
      this.scrollTop = y
    }
  }
}
