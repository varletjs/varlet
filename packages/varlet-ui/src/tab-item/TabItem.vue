<template>
  <var-swipe-item :class="classes(n(), [!current, n('--inactive')])" var-tab-item-cover>
    <slot v-if="initSlot" />
  </var-swipe-item>
</template>

<script lang="ts">
import VarSwipeItem from '../swipe-item'
import { defineComponent, ref, computed } from 'vue'
import { useTabsItems } from './provide'
import { props } from './props'
import type { Ref, ComputedRef } from 'vue'
import type { TabItemProvider } from './provide'
import { createNamespace } from '../utils/components'

const { n, classes } = createNamespace('tab-item')

export default defineComponent({
  name: 'VarTabItem',
  components: {
    VarSwipeItem,
  },
  props,
  setup(props) {
    const current: Ref<boolean> = ref(false)
    const initSlot: Ref<boolean> = ref(false)
    const name: ComputedRef<string | number | undefined> = computed(() => props.name)
    const { index, bindTabsItems } = useTabsItems()

    const setCurrent = (value: boolean) => {
      if (!initSlot.value && value) {
        initSlot.value = true
      }

      current.value = value
    }

    const tabItemProvider: TabItemProvider = {
      index,
      name,
      setCurrent,
    }

    bindTabsItems(tabItemProvider)

    return {
      n,
      classes,
      current,
      initSlot,
    }
  },
})
</script>

<style lang="less">
@import '../styles/common';
@import '../swipe-item/swipeItem';
@import './tabItem';
</style>
