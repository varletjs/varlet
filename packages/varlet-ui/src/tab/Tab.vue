<template>
  <div
    :class="classes(n(), 'var--box', computeColorClass(), n(`--${itemDirection}`))"
    ref="tabEl"
    v-ripple="{ disabled }"
    :style="{
      color: computeColorStyle(),
    }"
    @click="handleClick"
  >
    <slot />
  </div>
</template>

<script lang="ts">
import Ripple from '../ripple'
import { defineComponent, ref, computed, watch } from 'vue'
import { props } from './props'
import { createNamespace } from '../utils/components'
import { useTabs } from './provide'
import type { Ref, ComputedRef } from 'vue'
import type { TabProvider } from './provide'

const { n, classes } = createNamespace('tab')

export default defineComponent({
  name: 'VarTab',
  directives: { Ripple },
  props,
  setup(props) {
    const tabEl: Ref<HTMLElement | null> = ref(null)
    const name: ComputedRef<string | number | undefined> = computed(() => props.name)
    const disabled: ComputedRef<boolean> = computed(() => props.disabled)
    const element: ComputedRef<HTMLElement | null> = computed(() => tabEl.value)
    const { index, tabs, bindTabs } = useTabs()
    const { onTabClick, active, activeColor, inactiveColor, disabledColor, itemDirection, resize } = tabs

    const tabProvider: TabProvider = {
      name,
      index,
      disabled,
      element,
    }

    bindTabs(tabProvider)

    const computeColorStyle = () => {
      const { disabled, name } = props

      return disabled
        ? disabledColor.value
        : active.value === name || active.value === index?.value
        ? activeColor.value
        : inactiveColor.value
    }

    const computeColorClass = () => {
      const { disabled, name } = props

      return disabled
        ? 'var-tab--disabled'
        : active.value === name || active.value === index?.value
        ? 'var-tab--active'
        : 'var-tab--inactive'
    }

    const handleClick = (event: Event) => {
      const { disabled, name, onClick } = props

      if (disabled) {
        return
      }

      onClick?.(name ?? index.value, event)
      onTabClick(tabProvider)
    }

    watch(() => props.name, resize)
    watch(() => props.disabled, resize)

    return {
      n,
      classes,
      tabEl,
      active,
      activeColor,
      inactiveColor,
      itemDirection,
      computeColorStyle,
      computeColorClass,
      handleClick,
    }
  },
})
</script>

<style lang="less">
@import '../styles/common';
@import '../ripple/ripple';
@import './tab';
</style>
