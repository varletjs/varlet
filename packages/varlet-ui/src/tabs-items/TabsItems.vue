<template>
  <var-swipe :class="n()" ref="swipe" :loop="loop" :touchable="canSwipe" :indicator="false" @change="handleSwipeChange">
    <slot />
  </var-swipe>
</template>

<script lang="ts">
import VarSwipe from '../swipe'
import { defineComponent, watch, ref } from 'vue'
import { useTabItem } from './provide'
import { call, createNamespace } from '../utils/components'
import { props } from './props'
import type { Ref } from 'vue'
import type { TabsItemsProvider } from './provide'
import type { TabItemProvider } from '../tab-item/provide'

const { n } = createNamespace('tabs-items')

export default defineComponent({
  name: 'VarTabsItems',
  components: { VarSwipe },
  props,
  setup(props) {
    const swipe: Ref<null | typeof VarSwipe> = ref(null)
    const { tabItemList, bindTabItem, length } = useTabItem()

    const matchName = (active: number | string | undefined): TabItemProvider | undefined => {
      return tabItemList.find(({ name }: TabItemProvider) => active === name.value)
    }

    const matchIndex = (active: number | string | undefined): TabItemProvider | undefined => {
      return tabItemList.find(({ index }: TabItemProvider) => active === index.value)
    }

    const matchActive = (active: number | string | undefined): TabItemProvider | undefined => {
      return matchName(active) || matchIndex(active)
    }

    const handleActiveChange = (newValue: number | string | undefined) => {
      const newActiveTabItemProvider: TabItemProvider | undefined = matchActive(newValue)
      if (!newActiveTabItemProvider) {
        return
      }

      tabItemList.forEach(({ setCurrent }) => setCurrent(false))
      newActiveTabItemProvider.setCurrent(true)

      swipe.value?.to(newActiveTabItemProvider.index.value)
    }

    const handleSwipeChange = (currentIndex: number) => {
      const tabItem = tabItemList.find(({ index }) => index.value === currentIndex) as TabItemProvider
      const active = tabItem.name.value ?? tabItem.index.value

      call(props['onUpdate:active'], active)
    }

    // expose
    const getSwipe = () => {
      return swipe.value
    }

    const tabsItemsProvider: TabsItemsProvider = {}
    bindTabItem(tabsItemsProvider)

    watch(() => props.active, handleActiveChange)

    watch(
      () => length.value,
      () => handleActiveChange(props.active)
    )

    return {
      swipe,
      n,
      handleSwipeChange,
      getSwipe,
    }
  },
})
</script>

<style lang="less">
@import '../styles/common';
@import '../swipe/swipe';
</style>
