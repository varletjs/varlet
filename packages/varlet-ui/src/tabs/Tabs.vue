<template>
  <component
    :is="sticky ? n('$-sticky') : Transition"
    :ref="sticky ? 'stickyComponent' : undefined"
    :css-mode="sticky ? stickyCssMode : undefined"
    :offset-top="sticky ? offsetTop : undefined"
    :z-index="sticky ? stickyZIndex : undefined"
  >
    <div
      :class="
        classes(
          n(),
          n('$--box'),
          n(`--item-${itemDirection}`),
          n(`--layout-${layoutDirection}-padding`),
          formatElevation(elevation, 4),
          [fixedBottom, n('--fixed-bottom')],
          [safeArea, n('--safe-area')]
        )
      "
      :style="{ background: color }"
      v-bind="$attrs"
    >
      <div
        ref="scrollerEl"
        :class="
          classes(
            n('tab-wrap'),
            [localScrollable, n(`--layout-${layoutDirection}-scrollable`)],
            n(`--layout-${layoutDirection}`)
          )
        "
      >
        <slot />

        <div
          :class="classes(n('indicator'), n(`--layout-${layoutDirection}${indicatorPosition}-indicator`))"
          :style="{
            width: layoutDirection === 'horizontal' ? indicatorWidth : toSizeUnit(indicatorSize),
            height: layoutDirection === 'horizontal' ? toSizeUnit(indicatorSize) : indicatorHeight,
            transform: layoutDirection === 'horizontal' ? `translateX(${indicatorX})` : `translateY(${indicatorY})`,
            background: indicatorColor || activeColor,
          }"
        ></div>
      </div>
    </div>
  </component>
</template>

<script lang="ts">
import VarSticky from '../sticky'
import { defineComponent, watch, ref, computed, Transition, type Ref, type ComputedRef } from 'vue'
import { props } from './props'
import { useTabList, type TabsProvider } from './provide'
import { type TabProvider } from '../tab/provide'
import { isNumber } from '@varlet/shared'
import { linear } from '../utils/shared'
import { toSizeUnit, scrollTo, doubleRaf } from '../utils/elements'
import { createNamespace, call, formatElevation } from '../utils/components'
import { useEventListener } from '@varlet/use'

const { n, classes } = createNamespace('tabs')

export default defineComponent({
  name: 'VarTabs',
  components: { VarSticky },
  inheritAttrs: false,
  props,
  setup(props) {
    const indicatorWidth: Ref<string> = ref('0px')
    const indicatorHeight: Ref<string> = ref('0px')
    const indicatorX: Ref<string> = ref('0px')
    const indicatorY: Ref<string> = ref('0px')
    const localScrollable: Ref<boolean> = ref(false)
    const scrollerEl: Ref<HTMLElement | null> = ref(null)
    const active: ComputedRef<number | string> = computed(() => props.active)
    const indicatorPosition: ComputedRef<string> = computed(() =>
      props.indicatorPosition === 'reverse' ? '-reverse' : ''
    )
    const activeColor: ComputedRef<string | undefined> = computed(() => props.activeColor)
    const inactiveColor: ComputedRef<string | undefined> = computed(() => props.inactiveColor)
    const disabledColor: ComputedRef<string | undefined> = computed(() => props.disabledColor)
    const itemDirection: ComputedRef<string> = computed(() => props.itemDirection)
    const stickyComponent: Ref<null | typeof VarSticky> = ref(null)
    const { tabList, bindTabList, length } = useTabList()

    const onTabClick = (tab: TabProvider) => {
      const currentActive = tab.name.value ?? tab.index.value
      const { active, onChange, onClick } = props

      call(props['onUpdate:active'], currentActive)
      call(onClick, currentActive)
      currentActive !== active && call(onChange, currentActive)
    }

    const matchName = (): TabProvider | undefined => {
      return tabList.find(({ name }: TabProvider) => props.active === name.value)
    }

    const matchIndex = (activeIndex?: number): TabProvider | undefined => {
      return tabList.find(({ index }: TabProvider) => (activeIndex ?? props.active) === index.value)
    }

    const matchBoundary = (): TabProvider | undefined => {
      if (length.value === 0) {
        return
      }

      const { active } = props

      if (isNumber(active)) {
        const activeIndex = active > length.value - 1 ? length.value - 1 : 0
        call(props['onUpdate:active'], activeIndex)
        return matchIndex(activeIndex)
      }
    }

    const watchScrollable = () => {
      localScrollable.value = props.scrollable === 'always' || tabList.length >= 5
    }

    const moveIndicator = ({ element }: TabProvider) => {
      const el = element.value

      if (!el) return

      if (props.layoutDirection === 'horizontal') {
        indicatorWidth.value = `${el.offsetWidth}px`
        indicatorX.value = `${el.offsetLeft}px`
      } else {
        indicatorHeight.value = `${el.offsetHeight}px`
        indicatorY.value = `${el.offsetTop}px`
      }
    }

    const scrollToCenter = ({ element }: TabProvider) => {
      if (!localScrollable.value) {
        return
      }

      const scroller: HTMLElement = scrollerEl.value as HTMLElement
      const el = element.value as HTMLElement

      if (props.layoutDirection === 'horizontal') {
        const left: number = el.offsetLeft + el.offsetWidth / 2 - scroller.offsetWidth / 2
        scrollTo(scroller, {
          left,
          animation: linear,
        })
      } else {
        const top: number = el.offsetTop + el.offsetHeight / 2 - scroller.offsetHeight / 2
        scrollTo(scroller, {
          top,
          animation: linear,
        })
      }
    }

    // expose
    const resize = () => {
      const tab: TabProvider | undefined = matchName() || matchIndex() || matchBoundary()
      if (!tab || tab.disabled.value) {
        return
      }

      watchScrollable()
      moveIndicator(tab)
      scrollToCenter(tab)
    }

    // expose
    const resizeSticky = async () => {
      if (props.sticky && stickyComponent.value) {
        await stickyComponent.value.resize()
      }
    }

    const tabsProvider: TabsProvider = {
      active,
      activeColor,
      inactiveColor,
      disabledColor,
      itemDirection,
      resize,
      onTabClick,
    }

    bindTabList(tabsProvider)

    watch(
      () => length.value,
      async () => {
        await doubleRaf()
        resize()
      }
    )

    watch(() => props.active, resize)
    watch(() => props.scrollable, resize)
    useEventListener(window, 'resize', resize)

    return {
      stickyComponent,
      indicatorWidth,
      indicatorHeight,
      indicatorX,
      indicatorY,
      indicatorPosition,
      localScrollable,
      scrollerEl,
      Transition,
      toSizeUnit,
      n,
      classes,
      resize,
      resizeSticky,
      formatElevation,
    }
  },
})
</script>

<style lang="less">
@import '../styles/common';
@import '../styles/elevation';
@import '../sticky/sticky';
@import './tabs';
</style>
