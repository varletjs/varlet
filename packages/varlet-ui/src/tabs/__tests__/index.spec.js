import Tabs from '..'
import Tab from '../../tab'
import TabsItems from '../../tabs-items'
import TabItem from '../../tab-item'
import VarTabs from '../Tabs'
import VarTab from '../../tab/Tab'
import VarTabsItems from '../../tabs-items/TabsItems'
import VarTabItem from '../../tab-item/TabItem'
import { mount } from '@vue/test-utils'
import { createApp } from 'vue'
import { delay, mockOffset } from '../../utils/test'

let originScrollTo

beforeEach(() => {
  originScrollTo = Element.prototype.scrollTo
  Element.prototype.scrollTo = vi.fn()
})

afterEach(() => {
  Element.prototype.scrollTo = originScrollTo
})

mockOffset()

const Wrapper = {
  components: {
    [VarTabs.name]: VarTabs,
    [VarTab.name]: VarTab,
    [VarTabsItems.name]: VarTabsItems,
    [VarTabItem.name]: VarTabItem,
  },
  props: [
    'onClick',
    'onChange',
    'layoutDirection',
    'itemDirection',
    'fixedBottom',
    'color',
    'activeColor',
    'inactiveColor',
    'disabledColor',
    'indicatorColor',
    'indicatorSize',
    'elevation',
    'safeArea',
  ],
  data: () => ({
    active: 2,
  }),
  template: `
    <var-tabs v-model:active="active" v-bind="$props">
      <var-tab style="width: 100px">tab 1</var-tab>
      <var-tab disabled style="width: 100px">tab 2</var-tab>
      <var-tab style="width: 100px">tab 3</var-tab>
    </var-tabs>

    <var-tabs-items>
      <var-tab-item>tab item 1</var-tab-item>
      <var-tab-item>tab item 2</var-tab-item>
      <var-tab-item>tab item 3</var-tab-item>
    </var-tabs-items>
  `,
}

test('test tabs & tab & tabsItems & tabItem plugin', () => {
  const app = createApp({}).use(Tabs).use(Tab).use(TabsItems).use(TabItem)
  expect(app.component(Tabs.name)).toBeTruthy()
  expect(app.component(Tab.name)).toBeTruthy()
  expect(app.component(TabsItems.name)).toBeTruthy()
  expect(app.component(TabItem.name)).toBeTruthy()
})

test('test tabs event', async () => {
  const onClick = vi.fn()
  const onChange = vi.fn()
  const wrapper = mount(Wrapper, {
    props: {
      onClick,
      onChange,
    },
  })

  await delay(100)
  await wrapper.find('.var-tab').trigger('click')
  expect(onClick).toHaveBeenLastCalledWith(0)
  expect(onChange).toHaveBeenLastCalledWith(0)
  wrapper.unmount()
})

describe('test tabs component props', () => {
  test('test tabs layout-direction', async () => {
    const wrapper = mount(Wrapper, {
      props: {
        layoutDirection: 'horizontal',
      },
    })

    expect(wrapper.find('.var-tabs--layout-horizontal').exists()).toBe(true)
    await wrapper.setProps({ layoutDirection: 'vertical' })
    expect(wrapper.find('.var-tabs--layout-vertical').exists()).toBe(true)
    wrapper.unmount()
  })

  test('test tabs item-direction', async () => {
    const wrapper = mount(Wrapper, {
      props: {
        itemDirection: 'horizontal',
      },
    })

    expect(wrapper.find('.var-tabs--item-horizontal').exists()).toBe(true)
    await wrapper.setProps({ itemDirection: 'vertical' })
    expect(wrapper.find('.var-tabs--item-vertical').exists()).toBe(true)
    wrapper.unmount()
  })

  test('test tabs fixed-bottom', async () => {
    const wrapper = mount(Wrapper, {
      props: {
        fixedBottom: true,
      },
    })

    expect(wrapper.find('.var-tabs--fixed-bottom').exists()).toBe(true)
    await wrapper.setProps({ fixedBottom: false })
    expect(wrapper.find('.var-tabs--fixed-bottom').exists()).toBe(false)
    wrapper.unmount()
  })

  test('test tabs color', () => {
    const wrapper = mount(Wrapper, {
      props: {
        color: 'red',
      },
    })

    expect(wrapper.find('.var-tabs').attributes('style')).toContain('background: red;')
    wrapper.unmount()
  })

  test('test tabs active-color', async () => {
    const wrapper = mount(Wrapper, {
      props: {
        activeColor: 'red',
      },
    })

    await delay(100)
    expect(wrapper.find('.var-tab--active').attributes('style')).toContain('color: red;')
    wrapper.unmount()
  })

  test('test tabs inactive-color', async () => {
    const wrapper = mount(Wrapper, {
      props: {
        inactiveColor: 'red',
      },
    })

    await delay(100)
    expect(wrapper.find('.var-tab--inactive').attributes('style')).toContain('color: red;')
    wrapper.unmount()
  })

  test('test tabs disabled-color', () => {
    const wrapper = mount(Wrapper, {
      props: {
        disabledColor: 'red',
      },
    })

    expect(wrapper.find('.var-tab--disabled').attributes('style')).toContain('color: red;')
    wrapper.unmount()
  })

  test('test tabs indicator-color', async () => {
    const wrapper = mount(Wrapper, {
      props: {
        indicatorColor: 'red',
      },
    })

    await delay(100)
    expect(wrapper.find('.var-tabs--layout-horizontal-indicator').attributes('style')).toContain('background: red;')
    wrapper.unmount()
  })

  test('test tabs indicator-size', async () => {
    const wrapper = mount(Wrapper, {
      props: {
        indicatorSize: '10px',
      },
    })

    expect(wrapper.find('.var-tabs--layout-horizontal-indicator').attributes('style')).toContain('height: 10px;')
    await wrapper.setProps({ indicatorSize: 4 })
    expect(wrapper.find('.var-tabs--layout-horizontal-indicator').attributes('style')).toContain('height: 4px;')
    wrapper.unmount()
  })

  test('test tabs elevation', async () => {
    const wrapper = mount(Wrapper, {
      props: {
        elevation: true,
      },
    })

    expect(wrapper.find('.var-elevation--4').exists()).toBe(true)
    await wrapper.setProps({ elevation: false })
    expect(wrapper.find('.var-elevation--4').exists()).toBe(false)
    wrapper.unmount()
  })

  test('test tabs safe-area', async () => {
    const wrapper = mount(Wrapper, {
      props: {
        safeArea: true,
      },
    })

    expect(wrapper.find('.var-tabs--safe-area').exists()).toBe(true)
    await wrapper.setProps({ safeArea: false })
    expect(wrapper.find('.var-tabs--safe-area').exists()).toBe(false)
    wrapper.unmount()
  })
})
