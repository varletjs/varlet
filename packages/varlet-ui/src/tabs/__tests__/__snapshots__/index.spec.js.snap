// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test tabs active 1`] = `
"<transition-stub>
  <div class=\\"var-tabs var--box var-tabs--item-horizontal var-tabs--layout-horizontal-padding\\">
    <div class=\\"var-tabs__tab-wrap var-tabs--layout-horizontal\\">
      <div class=\\"var-tab var--box var-tab--inactive var-tab--horizontal\\" style=\\"width: 100px;\\">选项1</div>
      <div class=\\"var-tab var--box var-tab--inactive var-tab--horizontal\\" style=\\"width: 100px;\\">选项2</div>
      <div class=\\"var-tab var--box var-tab--active var-tab--horizontal\\" style=\\"width: 100px;\\">选项3</div>
      <div class=\\"var-tabs__indicator var-tabs--layout-horizontal-indicator\\" style=\\"width: 100px; transform: translateX(0px);\\"></div>
    </div>
  </div>
</transition-stub>
<div class=\\"var-swipe var-tabs-items\\">
  <div class=\\"var-swipe__track\\" style=\\"width: 0px; transform: translateX(0px); transition-duration: 300ms;\\">
    <div class=\\"var-swipe-item var-tab-item var-tab-item--inactive\\" style=\\"width: 0px; transform: translateX(0px);\\" var-tab-item-cover=\\"\\"></div>
    <div class=\\"var-swipe-item var-tab-item var-tab-item--inactive\\" style=\\"width: 0px; transform: translateX(0px);\\" var-tab-item-cover=\\"\\"></div>
    <div class=\\"var-swipe-item var-tab-item\\" style=\\"width: 0px; transform: translateX(0px);\\" var-tab-item-cover=\\"\\">视图3</div>
  </div>
  <!--v-if-->
</div>"
`;

exports[`test tabs disabled 1`] = `
"<transition-stub>
  <div class=\\"var-tabs var--box var-tabs--item-horizontal var-tabs--layout-horizontal-padding\\" modelvalue=\\"1\\">
    <div class=\\"var-tabs__tab-wrap var-tabs--layout-horizontal\\">
      <div class=\\"var-tab var--box var-tab--disabled var-tab--horizontal\\">选项1</div>
      <div class=\\"var-tab var--box var-tab--inactive var-tab--horizontal\\">选项2</div>
      <div class=\\"var-tab var--box var-tab--inactive var-tab--horizontal\\">选项3</div>
      <div class=\\"var-tabs__indicator var-tabs--layout-horizontal-indicator\\" style=\\"width: 0px; transform: translateX(0px);\\"></div>
    </div>
  </div>
</transition-stub>"
`;

exports[`test tabs relation tabsItems 1`] = `
"<transition-stub>
  <div class=\\"var-tabs var--box var-tabs--item-horizontal var-tabs--layout-horizontal-padding\\">
    <div class=\\"var-tabs__tab-wrap var-tabs--layout-horizontal\\">
      <div class=\\"var-tab var--box var-tab--inactive var-tab--horizontal\\" style=\\"width: 100px;\\">选项1</div>
      <div class=\\"var-tab var--box var-tab--inactive var-tab--horizontal\\" style=\\"width: 100px;\\">选项2</div>
      <div class=\\"var-tab var--box var-tab--active var-tab--horizontal\\" style=\\"width: 100px;\\">选项3</div>
      <div class=\\"var-tabs__indicator var-tabs--layout-horizontal-indicator\\" style=\\"width: 100px; transform: translateX(0px);\\"></div>
    </div>
  </div>
</transition-stub>
<div class=\\"var-swipe var-tabs-items\\">
  <div class=\\"var-swipe__track\\" style=\\"width: 0px; transform: translateX(0px); transition-duration: 300ms;\\">
    <div class=\\"var-swipe-item var-tab-item var-tab-item--inactive\\" style=\\"width: 0px; transform: translateX(0px);\\" var-tab-item-cover=\\"\\"></div>
    <div class=\\"var-swipe-item var-tab-item var-tab-item--inactive\\" style=\\"width: 0px; transform: translateX(0px);\\" var-tab-item-cover=\\"\\"></div>
    <div class=\\"var-swipe-item var-tab-item\\" style=\\"width: 0px; transform: translateX(0px);\\" var-tab-item-cover=\\"\\">视图3</div>
  </div>
  <!--v-if-->
</div>"
`;

exports[`test tabs relation tabsItems 2`] = `
"<transition-stub>
  <div class=\\"var-tabs var--box var-tabs--item-horizontal var-tabs--layout-horizontal-padding\\">
    <div class=\\"var-tabs__tab-wrap var-tabs--layout-horizontal\\">
      <div class=\\"var-tab var--box var-tab--inactive var-tab--horizontal\\" style=\\"width: 100px;\\">选项1</div>
      <div class=\\"var-tab var--box var-tab--active var-tab--horizontal\\" style=\\"width: 100px;\\">选项2</div>
      <div class=\\"var-tab var--box var-tab--inactive var-tab--horizontal\\" style=\\"width: 100px;\\">选项3</div>
      <div class=\\"var-tabs__indicator var-tabs--layout-horizontal-indicator\\" style=\\"width: 100px; transform: translateX(0px);\\"></div>
    </div>
  </div>
</transition-stub>
<div class=\\"var-swipe var-tabs-items\\">
  <div class=\\"var-swipe__track\\" style=\\"width: 0px; transform: translateX(0px); transition-duration: 300ms;\\">
    <div class=\\"var-swipe-item var-tab-item var-tab-item--inactive\\" style=\\"width: 0px; transform: translateX(0px);\\" var-tab-item-cover=\\"\\"></div>
    <div class=\\"var-swipe-item var-tab-item\\" style=\\"width: 0px; transform: translateX(0px);\\" var-tab-item-cover=\\"\\">视图2</div>
    <div class=\\"var-swipe-item var-tab-item var-tab-item--inactive\\" style=\\"width: 0px; transform: translateX(0px);\\" var-tab-item-cover=\\"\\">视图3</div>
  </div>
  <!--v-if-->
</div>"
`;
